### PR TITLE
Fix leak related to stop iteration

### LINKED ISSUE #1074: Memory Leak when using nn.Module hooks and thunder.jit
Using nn.Module hook with thunder.jit leads to memory leak. Disabling the hook, doesn't leak the memory.

This affects usage with Megatron-LM's DistributedOptimizer which uses hooks - https://github.com/NVIDIA/Megatron-LM/blob/73ca2e34ab58e65191a1329189720e1520eef263/megatron/core/optimizer/distrib_optimizer.py#L535-L542

Repro - 
```python
import weakref
import torch
import thunder
import refcycle  # pip install refcycle

class Identity(torch.nn.Module):
    def forward(self, x):
        return x

def main():
    with torch.device("cuda"):
        # model = torch.nn.Sequential(Identity())  # PyTorch eager doesn't leak

        model = torch.nn.Sequential(thunder.jit(Identity()))
        # More tensors the model has, more it leaks
        # model = torch.nn.Sequential(torch.nn.Linear(16, 16), thunder.jit(Identity()), torch.nn.Linear(16, 16))
        
        def my_noop_hook(module, unsued):
            for p in module.parameters():
                pass

        # Disabling the hook -> doesn't leak.
        remove_handle = torch.nn.modules.module.register_module_forward_pre_hook(my_noop_hook)
        x = torch.randn(16, 16)

    model(x)
    return weakref.ref(x)

weak_x = main()

snap = refcycle.snapshot()

for t in snap.find_by(lambda t: isinstance(t, torch.Tensor)):
    print(t.shape)  # Prints - torch.Size([16, 16])

# Code to find and save cycles
# for idx, anc in enumerate(snap.ancestors(weak_x())):
#     print(idx)
#     try:
#         cycle = snap.shortest_cycle(anc)
#         # cycle.export_image(f"cycles/cycle_{idx}.png")
#         print(cycle)
#     except:
#         pass
```

There are a lot of cycles but I am hoping resolving this first one might take care of others.

cycle_120

I could check the cycle in debugger with
```python
anc.f_locals["python_frame"].f_back.f_back is anc
```

And the path from frame (which owns the tensor and is part of the cycle) to the tensor is

path

```python
anc.f_locals["runtimectx"]._original_callsite[0].f_locals["args"][0]
```

@t-vi do you know where this could be coming from?

cc: @kiya00 @IvanYashchuk 

cc @apaz-cli

### PR DESCRIPTION
**Pull request by lantiga**
## What does this PR do?

We forgot to clean up the exception after handling `StopIteration` in the `for_iter` lookaside. This PR fixes it.
It also adds two tests for testing exception-related leaks.

Fixes #1074 

## Limitations

We still leak when we raise from inside a jitted function and catch on the outside. I'll open a separate issue, but this is unrelated to #1074

### CODE REVIEW

File: thunder/tests/test_interpreter.py [hunk 1]
```diff
@@ -779,6 +779,53 @@ def cross_function_exceptions():
...
+
+def test_exception_no_leak(jit):
+
+    class Identity(torch.nn.Module):
+        def raises():
+            raise RuntimeError("Exc")
+
+        def forward(self, x):
+            try:
+                raises()
```
Comment (t-vi) [inline thunder/tests/test_interpreter.py:811]:
```suggestion
                self.raises()
```
?

File: thunder/tests/test_interpreter.py [hunk 2]
```diff
@@ -779,6 +779,53 @@ def cross_function_exceptions():
...
+        model = thunder.jit(Identity())
+        x = torch.randn(16, 16)
+
+        model(x)
+
+        return weakref.ref(x)
+
+    weak_x = foo()
+
+    assert weak_x is None
```
Comment (t-vi) [inline thunder/tests/test_interpreter.py:800]:
```suggestion
    assert weak_x() is None
```
?

File: thunder/tests/test_interpreter.py [hunk 3]
```diff
@@ -779,6 +779,53 @@ def cross_function_exceptions():
...
+
+    weak_x = foo()
+
+    assert weak_x() is None
+
+
+def test_exception_no_leak(jit):
+
+    class Identity(torch.nn.Module):
+        def raises():
```
Comment (t-vi) [inline thunder/tests/test_interpreter.py:806]:
```suggestion
        @staticmethod
        def raises():
```

File: thunder/core/interpreter.py [hunk 4]
```diff
@@ -1175,40 +1175,55 @@ def get_localsplus_name(self, idx: int) -> str:
...
+        # we run the getting of the frame in a separate thread because
+        # we want to avoid having f_back pointing to the function
+        # handling the error
+        import _thread
+
+        result_container = []
+        _thread.start_new_thread(get_frame, (result_container, rel_lineno, filename, firstlineno, name, qualname))
+        while not result_container:
+            pass
+        return result_container[0]
```
Comment (t-vi) [inline thunder/core/interpreter.py:1226]: So this is needed to remove this cycle (image by @kshitij12345 ) that apparently causes trouble for us.
Before this, I could reproduce the problem by running the (`-k leak`) tests ~5-10 times, not I can run 50 iterations without anything happening.

image

Comment (lantiga) [reply thunder/core/interpreter.py:1226]: Interesting, this may solve the cross-boundary exception leak too

**Pull request merged**